File: vkatz-lib/src/main/java/by/vkatz/utils/SelfKillerExecutor.java
package by.vkatz.utils;

import java.util.Stack;
import java.util.concurrent.Executor;

/**
 * Created by vKatz on 27.02.2015.
 */
@SuppressWarnings("unused")
public class SelfKillerExecutor implements Executor {
    private final Object sync = new Object();
    private Thread thread = null;
    private Stack<Runnable> tasks = new Stack<>();

    @Override
    public void execute(Runnable runnable) {
        tasks.push(runnable);
        synchronized (sync) {
            if (thread == null) {
                thread = new Thread(new Runnable() {
                    @Override
                    public void run() {
                        while (!tasks.isEmpty()) try {
                            tasks.pop().run();
                        } catch (Exception e) {
                            e.printStackTrace();
                        }
                        synchronized (sync) {
                            thread = null;
                        }
                    }
                });
                thread.start();
            }
        }
    }
}

File: vkatz-lib/src/main/kotlin/by/vkatz/utils/SelfKillerExecutor.kt
package by.vkatz.utils

import java.util.*
import java.util.concurrent.Executor

class SelfKillerExecutor : Executor {
    private val sync = Any()
    private var thread: Thread? = null
    private val tasks = Stack<Runnable>()

    override fun execute(runnable: Runnable) {
        tasks.push(runnable)
        synchronized(sync) {
            if (thread == null) {
                thread = Thread {
                    while (!tasks.isEmpty())
                        try {
                            tasks.pop().run()
                        } catch (e: Exception) {
                            e.printStackTrace()
                        }
                    synchronized(sync) {
                        thread = null
                    }
                }
                thread!!.start()
            }
        }
    }
}